scrapy/http/request/__init__.py

```diff
@@ -50,8 +50,8 @@ def _set_url(self, url):
         if not isinstance(url, six.string_types):
             raise TypeError('Request url must be str or unicode, got %s:' % type(url).__name__)
 
-        url = to_native_str(url, self.encoding)
-        self._url = escape_ajax(safe_url_string(url))
+        s = safe_url_string(url, self.encoding)
```

Comment:
Aha, so let's say in Python 2 a unicode URL is extracted from cp1251 page and the encoding is not passed. Before the change:
 
* url is encoded to utf8,
* then escaped. It means path is escaped correctly, but query arguments and fragment are not.

Now:
* path is escaped correctly, query and fragments are not.

The same, but with bytes URL. Before:

* url is passed to safe_url_string as-is; it means path is escaped incorrectly, but query and fragment are escaped correctly.

After:

* safe_url_string tries to convert URL to unicode. Sometimes it fails; w3lib escapes failing parts. For escaped parts path is encoded incorrectly, but query and fragment are encoded properly. For unescaped parts the whole URL is escaped incorrectly because unicode value is decoded from a wrong encoding.

So you're right, either way something is not correct if encoding is not passed. It is a bit worse IMHO after the change for binary URLs (less predictable), but either way it is broken, so the behavior change looks good to me.